### PR TITLE
Add another unique identifier for template cachekey (#23751)

### LINKED ISSUE #23751: Better way to cache template path in wc_get_template_part and wc_get_template  function that will work in a load balanced environment
### DETAILS OF THE ISSUE
We have a setup of multiple application servers behind a load balancer where there are different root path for each app server e.g.
Server 1 path - /svr/bindings/serverpathxxxxxxxxx1/
Server 2 path - /svr/bindings/serverpathxxxxxxxxx2/
And we have a single persistent Redis cache that serves those application containers

Based from the [`wc_get_template_part`](https://github.com/woocommerce/woocommerce/blob/master/includes/wc-core-functions.php#L159-L197) function, it stores the full path in the cache in the $template variable eg. /svr/bindings/serverpathxxxxxxxxx1/wp-content/plugins/woocommerce

Similar code happens in `wc_get_template` function in same file.

What happens in a load balanced environment with a single persistent cache server serving multiple server, where full paths on each server are different, is when the event happens that the **2nd server** is called and the cache for the $template is called for **1st server** resulting a `failed to open stream: No such file or directory in /svr/bindings/serverpathxxxxxxxxx1/wp-content/plugins/woocommerce` warning and not rendering all relevant WooCommerce templates.

### Possible resolutions
We are looking to discuss possible solutions and would love to try to submit a PR for the possible fix but would like to hear what would be feasible that would benefit similar users that are having the same setup and issues :
1) Caching relative path instead of the full path
2) Store 2 or more cache key entries for $template variable depending which server is calling 
3) Adding another checking mechanism for the $template path if it exists, then get the correct 

### Temporary workaround
At the moment for the people that are being blocked by this issue, we have seen workarounds like:
1) Skip the whole WooCommerce caching by adding this line of code `wp_cache_add_non_persistent_groups( array( 'woocommerce' ) );` in their theme's function.php or their custom plugin
2) Doing a core hack by commenting out these lines https://github.com/woocommerce/woocommerce/blob/90aa716098cdfaa5f411b23bd294f13f6a9b4b57/includes/wc-core-functions.php#L161 
https://github.com/woocommerce/woocommerce/blob/90aa716098cdfaa5f411b23bd294f13f6a9b4b57/includes/wc-core-functions.php#L209
and replacing it with `$template = '';` so the values will not be loaded from cache
3) Add this filter can be added in custom mu-plugin, theme's functions.php or a custom plugin.
https://gist.github.com/carl-alberto/fc0496962e3ae3f514e4c7374db08c9e

### PR DESCRIPTION
**Pull request by carl-alberto**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

In relation to #23751, this PR includes the proposed solution `2` which will create a cache key for each server and using the server variable `$_SERVER['SERVER_NAME']` as another layer of unique identifier for the $cachekey as there are no other elegant way to do this via native functions like `WC()->template_path()` and `WC()->plugin_path()` because it will always still use the full server path

Closes #23751 .

### How to test the changes in this Pull Request:

1. Can be reproduced on servers that have multiple containers. ( Pantheon is an example and in Performance medium and above plans, I can share access) The setup of multiple application servers behind a load balancer where there are different root path for each app server e.g.
Server 1 path - /svr/bindings/serverpathxxxxxxxxx1/
Server 2 path - /svr/bindings/serverpathxxxxxxxxx2/

Based from the wc_get_template_part function, it stores the full path in the cache in the $template variable eg. /svr/bindings/serverpathxxxxxxxxx1/wp-content/plugins/woocommerce

Similar code happens in wc_get_template function in same file.

What happens in a load balanced environment, where full paths on each server are different, is when the event happens that the 2nd server is called and the cache for the $template is called for 1st server resulting a failed to open stream: No such file or directory in /svr/bindings/serverpathxxxxxxxxx1/wp-content/plugins/woocommerce warning and not rendering all relevant WooCommerce templates.

2. When the patch is applied, error encountered in the previous step is eliminated

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

Improved the naming of template path cache keys in `wc_get_template_part` and `wc_get_template` function so it can also properly cache file paths that are unique in each container in a load balanced environment

### CODE REVIEW

File: includes/wc-core-functions.php [hunk 1]
```diff
@@ -157,7 +157,12 @@ function wc_update_order( $args ) {
  * @param string $name Template name (default: '').
  */
 function wc_get_template_part( $slug, $name = '' ) {
-	$cache_key = sanitize_key( implode( '-', array( 'template-part', $slug, $name, WC_VERSION ) ) );
+	if ( isset( $_SERVER['SERVER_NAME'] ) ) {
+		$uniquepath = sanitize_key( $_SERVER['SERVER_NAME'] );
```
Comment (claudiosanches) [inline includes/wc-core-functions.php]: This variable name don't follow WP standards, should use `_` like `$unique_path`. And this should not help everyone, probably better have an filter here too.

```
$server_name = isset( $_SERVER['SERVER_NAME'] ) ? sanitize_key( $_SERVER['SERVER_NAME'] ) : '';

$cache_key = sanitize_key( apply_filters( 'woocommerce_get_template_part_cache_key', implode( '-', array( 'template-part', $slug, $name, WC_VERSION, $server_name ) ), $slug, $name, WC_VERSION, $server_name ) );
```

File: includes/wc-core-functions.php [hunk 2]
```diff
@@ -205,7 +210,12 @@ function wc_get_template_part( $slug, $name = '' ) {
  * @param string $default_path  Default path. (default: '').
  */
 function wc_get_template( $template_name, $args = array(), $template_path = '', $default_path = '' ) {
-	$cache_key = sanitize_key( implode( '-', array( 'template', $template_name, $template_path, $default_path, WC_VERSION ) ) );
+	if ( isset( $_SERVER['SERVER_NAME'] ) ) {
+		$uniquepath = sanitize_key( $_SERVER['SERVER_NAME'] );
+	} else {
+		$uniquepath = '';
+	}
+	$cache_key = sanitize_key( implode( '-', array( 'template', $template_name, $template_path, $default_path, WC_VERSION, $uniquepath ) ) );
```
Comment (claudiosanches) [inline includes/wc-core-functions.php]: Here we can do the same as I suggested before, but here the filter should be `woocommerce_get_template_cache_key`.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
This changes seems good, but still is nice to have a filter to help here if some more changes is required.

File: includes/wc-core-functions.php (hunk 1 above)
Comment (carl-alberto) [reply includes/wc-core-functions.php]: Thanks for the feedback @claudiosanches , updated my PR here https://github.com/woocommerce/woocommerce/pull/23783/commits/217a95e937b6a2b5e54b0e5ee1023c4ae6cc7946

File: includes/wc-core-functions.php (hunk 2 above)
Comment (carl-alberto) [reply includes/wc-core-functions.php]: Thanks for the feedback @claudiosanches , updated my PR here https://github.com/woocommerce/woocommerce/pull/23783/commits/217a95e937b6a2b5e54b0e5ee1023c4ae6cc7946 as per your advice

**Comment by kfoubert** [comment]
Thanks, you guys rock.

**Comment by Uriel1339** [comment]
Great work on this. The faster it can get merged the better. Without this fix various larger plans hosted by Pantheon (where our WooCommerce store lives) are simply breaking because WooCommerce doesn't know what is where.

Looking forwards to get it merged in the next WC version!

**Comment by markjaquith** [comment]
# This is not a good solution

1. It fragments the cache — now every server with a different `$_SERVER['SERVER_NAME']` is accessing a different cache item.
2. It makes the assumption that servers with different document roots also have different server names. This is not a valid assumption.

Let's back up and look at the problem.

The root issue is that full filesystem paths are being cached, and full filesystems paths are not static.

The solution to this problem is to **stop caching full filesystem paths**.

> ---
>
> ### Never store a web app's full filesystem paths in a database or a cache.
>
> ---

By storing these, you're tying all instances of the web app to a concept it should not know about or care about: where on one particular machine's filesystem in a specific moment in time the application was installed.

You already know where the WordPress app is installed. `ABSPATH`. That gets set on every page load, on every server. It is correct, because if it were not, WordPress would not work.

So...

# Better solution

1. Before you store it, remove `ABSPATH` from the front of the absolute path.
2. After you retrieve it, add `ABSPATH` to the front of the stored relative path.

**Comment by carl-alberto** [comment]
thanks for the feedback @markjaquith , let me revise my PR to reflect that suggestion then

**Comment by JulienMelissas** [comment]
@markjaquith -- thank you so much for your comment. I completely agree.

I just spent a few hours trying to figure out what was going on with our store (we serve from multiple containers) and found out that the absolute paths are being cached. That's def. not a good solution in the first place.

@carl-alberto, please let me know if you would like some help on this PR, as we are effected and would like to be able to upgrade. Or if you have started with this and would like another set of eyes on it, I'd love to help!

**Comment by JulienMelissas** [comment]
If anyone is looking at this issue and wondering if there is a quick fix, I'm using this as an `mu-plugin` in my staging environment and it is working well: https://gist.github.com/JulienMelissas/d3601af950221432a109c8c4b1617b87

**Comment by carl-alberto** [comment]
Hi @JulienMelissas, still stumped on how to approach this issue as there is no native way https://codex.wordpress.org/Determining_Plugin_and_Content_Directories
 to not include the ABSPATH from these functions:

wc_locate_template
WC()->plugin_path()
file_exists
locate_template

Any additional input on how to exclude the ABSPATH from the template paths is highly appreciated by everyone

**Comment by kloon** [comment]
Closing in favor of #23971, would appreciate some testing of that.

**Pull request closed without merging**